tox.ini

```diff
@@ -78,11 +80,14 @@ pass_env =
     CI
     GITHUB_OUTPUT
     SECRETS_FROM_GITHUB
+    LIBJUJU_VERSION_SPECIFIER
 allowlist_externals =
     {[testenv:pack-wrapper]allowlist_externals}
 commands_pre =
     poetry install --only integration
     {[testenv:pack-wrapper]commands_pre}
+    # Hacking in after locked deps won't up/downgrade
+    pip install juju=={env:LIBJUJU_VERSION_SPECIFIER:2.9.44.0}
```

Comment:
Yes, we're taking 2.9 from `poetry.lock`, and then we overwrite it with 3.1 (in case it was selected).

Doesn't seem very elegant but I couldn't find a way to have `poetry.lock` maintaining the logic of both versions. And given that hopefully we're dropping 2.9 one day, as a temporary solution it's... well... doing the job.

Reply:
VERY NICE thank U so much!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

I'm adding it to the PR